### PR TITLE
[ML] clear job size estimate cache when feature is reset

### PR DESCRIPTION
**Pull request by benwtrent**
Since the feature reset API clears out the `.ml-*` indices, it follows that it also deletes the machine learning jobs. 

But, since the regular path of calling the delete job API is not followed, jobs that no longer exist could still have memory estimates cached on the master node. These would never get cleared out until after a master node changed. 

This commit causes feature reset to: 
 - await for all refresh requests to finish (of which there should usually be NONE as all assignments have been cancelled)
 - clear out the cached hashmap of memory estimates sitting on the master node
 - Then once cleared, new refreshes are allowed again

### CODE REVIEW

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/MachineLearning.java
Comment (droberts195) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/MachineLearning.java:1368]: Since the memory tracker cleanup waits for refreshes to finish, I would do this index cleanup after clearing the memory tracker.  It should avoid logging of spurious errors from refreshes that fail because the indices they're accessing get deleted.

Or was there a good reason for clearing the memory tracker last?

Comment (benwtrent) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/MachineLearning.java:1368]: @droberts195, I figured if the job's potentially still existed, it would be good to keep around their estimates. But, since all jobs should be closed by this point, clearing the tracker earlier is probably ok.

Comment (droberts195) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/MachineLearning.java:1368]: Yes, the jobs should be closed, and banned from reopening by the reset-in-progress cluster setting.  I think flipping the order may avoid log spam.

**Review by droberts195 — APPROVED** [review]
LGTM if you can get to the bottom of the test failure

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/autoscaling/MlAutoscalingDeciderService.java
Comment (benwtrent) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/autoscaling/MlAutoscalingDeciderService.java:378]: @droberts195 the test failure was a VALID test failure :). The test was taking advantage of the fact that the ml tracker is actually update to date immediately. 

When running the test directly, this is fine as the master node boots immediately and thus the memory is fresh.

When running the test AFTER another test, the cluster stays up and the tracker is reset. 

So, to improve this behavior, I added this clause. It is a no-brainer really. Autoscaling should not really check anything if there are `0` ML tasks and should be unaffected by memory tracker staleness.

**Comment by benwtrent** [comment]
@droberts195 requesting re-review as the change to fix the test was non-trivial.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/autoscaling/MlAutoscalingDeciderService.java
Comment (davidkyle) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/autoscaling/MlAutoscalingDeciderService.java:387]:
```suggestion
                                "Passing currently perceived capacity as down scale delay has not been satisfied; configured delay [%s]"
```

**Review by droberts195 — APPROVED** [review]
LGTM if you could just fix the typo Dave K pointed out before merging

**Comment by benwtrent** [comment]
@elasticmachine update branch

**Pull request merged**
